bmt/bmt.go

```diff
@@ -0,0 +1,565 @@
+
+// returns the size
+func (self *BMTHasher) Size() int {
+	return self.size
+}
+
+// returns the block size
+func (self *BMTHasher) BlockSize() int {
+	return self.blocksize
+}
+
+// hash.Hash interface Sum method appends the byte slice to the underlying
+// data before it calculates and returns the hash of the chunk
+func (self *BMTHasher) Sum(b []byte) (r []byte) {
+	t := self.getTree()
+	self.Write(b)
+	i := self.cur
+	//fmt.Printf("finalise for node index %v (leaves: %v)\n", i, len(t.leaves))
+	n := t.leaves[i]
+	j := i
+	// must run strictly before all nodes calculate
+	// datanodes are guaranteed to have a parent
+	if len(self.segment) > self.size && i > 0 && n.parent != nil {
+		n = n.parent
+	} else {
+		i *= 2
+	}
+	d := self.finalise(n, i)
+	//fmt.Printf("finalise for node level %v index %v depth %v, %v, %v\n", n.level, i, d, j, len(self.segment))
+	self.writeSegment(j, self.segment, d)
+	c := <-self.result
+	self.releaseTree()
+	return c
+}
+
+// BMTHasher implements the io.Writer interface
+// Write fills the buffer to hash
+// with every full segment complete launches a hasher go routine
+// that shoots up the BMT
+func (self *BMTHasher) Write(b []byte) (int, error) {
+	l := len(b)
+	if l <= 0 {
+		return 0, nil
+	}
+	s := self.segment
+	i := self.cur
+	count := (self.count + 1) / 2
+	need := self.count*self.size - self.cur*2*self.size
+	size := self.size
+	if need > size {
+		size *= 2
+	}
+	if l < need {
+		need = l
+	}
+	// calculate missing bit to complete current open segment
+	rest := size - len(s)
+	if need < rest {
+		rest = need
+	}
+	s = append(s, b[:rest]...)
+	need -= rest
+	//fmt.Printf("l: %v, s: %x, need: %v, size: %v, index: %v\n", l, s, need, size, self.cur)
+	// read full segments and the last possibly partial segment
+	for need > 0 && i < count-1 {
+		// push all finished chunks we read
+		self.writeSegment(i, s, self.depth)
+		need -= size
+		if need < 0 {
+			size += need
+		}
+		s = b[rest : rest+size]
+		rest += size
+		i++
+	}
+	//fmt.Printf("open segment %v len %v\n (data: %v)\n", i, len(s), hashstr(s))
+	self.segment = s
+	self.cur = i
+	// otherwise, we can assume len(s) == 0, so all buffer is read and chunk is not yet full
+	return l, nil
+}
+
+// reads from io.Reader and appends to the data to hash using Write
+// it reads so that chunk to hash is maximum length or reader reaches if EOF
+func (self *BMTHasher) ReadFrom(r io.Reader) (m int64, err error) {
+	self.getTree()
+	bufsize := self.size*self.count - self.size*self.cur - len(self.segment)
+	buf := make([]byte, bufsize)
+	var read int
+	for {
+		var n int
+		n, err = r.Read(buf)
+		read += n
+		if err == io.EOF || read == len(buf) {
+			hash := self.Sum(buf[:n])
+			if read == len(buf) {
+				err = NewEOC(hash)
+			}
+			break
+		}
+		if err != nil {
+			break
+		}
+		n, err = self.Write(buf[:n])
+		if err != nil {
+			break
+		}
+	}
+	return int64(read), err
+}
+
+func (self *BMTHasher) Reset() {
+	self.releaseTree()
```

Comment:
the release should not be here, but in Sum 
There is not much point in releasing here if we immediately reclaim with getTree